README.md
Introduction
============

PXF is an extensible framework that allows a distributed database like GPDB to query external data files, whose metadata is not managed by the database.
PXF includes built-in connectors for accessing data that exists inside HDFS files, Hive tables, HBase tables and more.
Users can also create their own connectors to other data storages or processing engines.
To create these connectors using JAVA plugins, see the PXF API and Reference Guide onGPDB.


Package Contents
================

## server/
Contains the server side code of PXF along with the PXF Service and all the Plugins

## cli/
Contains command line interface code for PXF

## automation/
Contains the automation and integration tests for PXF against the various datasources

## singlecluster/
Hadoop testing environment to exercise the pxf automation tests

## concourse/
Resources for PXF's Continuous Integration pipelines


PXF Development
=================
Below are the steps to build and install PXF along with its dependencies including GPDB and Hadoop.

To start, ensure you have a `~/workspace` directory and have cloned the `pxf` and its prerequisites(shown below) under it.
(The name `workspace` is not strictly required but will be used throughout this guide.)
```bash
mkdir -p ~/workspace
cd ~/workspace

git clone https://github.com/greenplum-db/pxf.git
```
Alternatively, you may create a symlink to your existing repo folder.
```bash
ln -s ~/<git_repos_root> ~/workspace
```

## Install Dependencies

To build PXF, you must have:

- JDK 1.8 (6/7/2019: can't go above 1.8 or it won't work)
- Go (1.9 or later)

To install Go on CentOS, `sudo yum install go`.

For other platforms, see the [Go downloads page](https://golang.org/dl/).

Once you have installed Go, you will need the `dep` and `ginkgo` tools, which install Go dependencies and run Go tests,
respectively. Assuming `go` is on your `PATH`, you can run:

```
go get github.com/golang/dep/cmd/dep
go get github.com/onsi/ginkgo/ginkgo
```

to install them.

## How to Build
PXF uses gradle for build and has a wrapper makefile for abstraction
```bash
cd ~/workspace/pxf

# Compile & Test PXF
make
  
# Simply Run unittest
make unittest
```

## Demonstrating Hadoop Integration
In order to demonstrate end to end functionality you will need GPDB and Hadoop installed.

### Hadoop
We have all the related hadoop components (hdfs, hive, hbase, zookeeper, etc) mapped into simple artifact named singlecluster.
You can [download from here](http://storage.googleapis.com/pxf-public/singlecluster-HDP.tar.gz) and untar the `singlecluster-HDP.tar.gz` file, which contains everything needed to run Hadoop.

```bash
mv singlecluster-HDP.tar.gz ~/workspace/
cd ~/workspace
tar xzf singlecluster-HDP.tar.gz
```

### GPDB
```
git clone https://github.com/greenplum-db/gpdb.git
```

You'll end up with a directory structure like this:

```
~
└── workspace
    ├── pxf
    ├── singlecluster-HDP
    └── gpdb
```

If you already have GPDB installed and running using the instructions shown in the [GPDB README](https://github.com/greenplum-db/gpdb), 
you can ignore the `Setup GPDB` section below and simply follow the steps in  `Setup Hadoop` and `Setup PXF`

If you don't wish to use docker, make sure you manually install JDK.

## Development With Docker
NOTE: Since the docker container will house all Single cluster Hadoop, Greenplum and PXF, we recommend that you have at least 4 cpus and 6GB memory allocated to Docker. These settings are available under docker preferences.

The following commands run the docker container and set up and switch to user gpadmin.

```bash
# Get the latest image
docker pull pivotaldata/gpdb-pxf-dev:centos6

# If you want to use gdb to debug gpdb you need the --privileged flag in the command below
docker run --rm -it \
  -p 5432:5432 \
  -p 5888:5888 \
  -p 8000:8000 \
  -p 5005:5005 \
  -p 8020:8020 \
  -p 9000:9000 \
  -p 9090:9090 \
  -p 50070:50070 \
  -w /home/gpadmin/workspace \
  -v ~/workspace/gpdb:/home/gpadmin/workspace/gpdb \
  -v ~/workspace/pxf:/home/gpadmin/workspace/pxf \
  -v ~/workspace/singlecluster-HDP:/home/gpadmin/workspace/singlecluster \
  pivotaldata/gpdb-pxf-dev:centos6 /bin/bash -c \
  "/home/gpadmin/workspace/pxf/dev/set_up_gpadmin_user.bash && /sbin/service sshd start && su - gpadmin"
```

### Setup GPDB

Configure, build and install GPDB. This will be needed only when you use the container for the first time with GPDB source.
```bash
~/workspace/pxf/dev/build_gpdb.bash
~/workspace/pxf/dev/install_gpdb.bash
```

For subsequent minor changes to GPDB source you can simply do the following:
```bash
~/workspace/pxf/dev/install_gpdb.bash
```

Run all the instructions below and run GROUP=smoke (in one script):
```bash
~/workspace/pxf/dev/smoke_shortcut.sh
```

Create Greenplum Cluster
```bash
source /usr/local/greenplum-db-devel/greenplum_path.sh
make -C ~/workspace/gpdb create-demo-cluster
source ~/workspace/gpdb/gpAux/gpdemo/gpdemo-env.sh
```

### Setup Hadoop
Hdfs will be needed to demonstrate functionality. You can choose to start additional hadoop components (hive/hbase) if you need them.

Setup [User Impersonation](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/Superusers.html) prior to starting the hadoop components (this allows the `gpadmin` user to access hadoop data).
```bash
~/workspace/pxf/dev/configure_singlecluster.bash
```

Setup and start HDFS
```bash
pushd ~/workspace/singlecluster/bin
echo y | ./init-gphd.sh
./start-hdfs.sh
popd
```

Start other optional components based on your need
```bash
pushd ~/workspace/singlecluster/bin
# Start Hive
./start-yarn.sh
./start-hive.sh

# Start HBase 
./start-zookeeper.sh
./start-hbase.sh
popd
```
### Setup Minio (optional)
Minio is an S3-API compatible local storage solution. The development docker image comes with Minio software pre-installed. To start the Minio server, run the following script:
```bash
source ~/workspace/pxf/dev/start_minio.bash
```
After the server starts, you can access Minio UI at `http://localhost:9000` from the host OS. Use `admin` for the access key and `password` for the secret key when connecting to your local Minio instance.

The script also sets `PROTOCOL=minio` so that the automation framework will use the local Minio server when running S3 automation tests. If later you would like to run Hadoop HDFS tests, unset this variable with `unset PROTOCOL` command.

### Setup PXF

Install PXF Server
```bash
# Install PXF
make -C ~/workspace/pxf install

# Initialize PXF
export PXF_CONF=~/pxf
export PXF_JVM_OPTS="-Xmx512m -Xms256m"
$PXF_HOME/bin/pxf init

# Start PXF
$PXF_HOME/bin/pxf start
```

Install PXF client (ignore if this is already done)
```bash
if [[ -d ~/workspace/gpdb/gpAux/extensions/pxf ]]; then
	PXF_EXTENSIONS_DIR=gpAux/extensions/pxf
else
	PXF_EXTENSIONS_DIR=gpcontrib/pxf
fi
make -C ~/workspace/gpdb/${PXF_EXTENSIONS_DIR} installcheck
psql -d template1 -c "create extension pxf"
```

### Run PXF Tests
All tests use a database named `pxfautomation`.
```bash
pushd ~/workspace/pxf/automation

# Initialize default server configs using template
cp ~/pxf/templates/*.xml ~/pxf/servers/default

# Run specific tests. Example: Hdfs Smoke Test
make TEST=HdfsSmokeTest

# Run all tests. This will be very time consuming.
make GROUP=gpdb

# If you wish to run test(s) against a different storage protocol set the following variable (for eg: s3) 
export PROTOCOL=s3
popd
```

If you see any HBase failures, try copying `pxf-hbase-*.jar` to the HBase classpath, and restart HBase:

```
cp ${PXF_HOME}/lib/pxf-hbase-*.jar ~/workspace/singlecluster/hbase/lib
~/workspace/singlecluster/bin/stop-hbase.sh
~/workspace/singlecluster/bin/start-hbase.sh
```

### Make Changes to PXF

To deploy your changes to PXF in the development environment.

```bash
# $PXF_HOME folder is replaced each time you make install.
# So, if you have any config changes, you may want to back those up.
$PXF_HOME/bin/pxf stop
make -C ~/workspace/pxf install
# Make any config changes you had backed up previously
rm -rf $PXF_HOME/pxf-service
yes | $PXF_HOME/bin/pxf init
$PXF_HOME/bin/pxf start
```

## IDE Setup (IntelliJ)

- Start IntelliJ. Click "Open" and select the directory to which you cloned the `pxf` repo.
- Select `File > Project Structure`.
- Make sure you have a JDK (version 1.8) selected.
- In the `Project Settings > Modules` section, import a Gradle module for the `pxf/server` directory. You may see an error saying that there's
no JDK set for Gradle. Just cancel and retry. It goes away the second time.
- In the `Project Settings > Modules` section, import a Maven module for the `pxf/automation` directory. You may do this by selecting the `pxf/automation/pom.xml` file.
- Restart IntelliJ
- Check that it worked by running a unit test (cannot currently run automation tests from IntelliJ) and making sure that imports, variables, and auto-completion function in the two modules.
- Optionally you can replace `${PXF_TMP_DIR}` with `${GPHOME}/pxf/tmp` in `automation/pom.xml`
- Select `Tools > Create Command-line Launcher...` to enable starting Intellij with the `idea` command, e.g. `cd ~/workspace/pxf && idea .`.

# To run a Kerberized Hadoop Cluster

## Requirements

- Download bin_gpdb (from any of the pipelines)
- Download pxf_tarball (from any of the pipelines)

These instructions allow you to run a Kerberized cluster

```bash
docker run --rm -it \
  --privileged \
  --hostname c6401.ambari.apache.org \
  -p 5432:5432 \
  -p 5888:5888 \
  -p 8000:8000 \
  -p 8080:8080 \
  -p 8020:8020 \
  -p 9000:9000 \
  -p 9090:9090 \
  -p 50070:50070 \
  -w /home/gpadmin/workspace \
  -v ~/workspace/gpdb:/home/gpadmin/workspace/gpdb_src \
  -v ~/workspace/pxf:/home/gpadmin/workspace/pxf_src \
  -v ~/workspace/singlecluster-HDP:/home/gpadmin/workspace/singlecluster \
  -v ~/Downloads/bin_gpdb:/home/gpadmin/workspace/bin_gpdb \
  -v ~/Downloads/pxf_tarball:/home/gpadmin/workspace/pxf_tarball \
  -e CLUSTER_NAME=hdp \
  -e NODE=c6401.ambari.apache.org \
  -e REALM=AMBARI.APACHE.ORG \
  -e TARGET_OS=centos \
  pivotaldata/gpdb-pxf-dev:centos6-hdp-secure /bin/bash

# Inside the container run the following command:
pxf_src/concourse/scripts/test_pxf_secure.bash

echo "+----------------------------------------------+"
echo "| Kerberos admin principal: admin/admin@$REALM |"
echo "| Kerberos admin password : admin              |"
echo "+----------------------------------------------+"

su - gpadmin
```


CHANGELOG.md
# Changelog

## 5.10.1 (01/07/2020)

#### Enhancements:

- [#276](https://github.com/greenplum-db/pxf/pull/276) Refactor filter parser code (Non-user facing)

#### Bug Fixes:

- [#283](https://github.com/greenplum-db/pxf/pull/283) Fix parquet write decimal
- [#280](https://github.com/greenplum-db/pxf/pull/280) Introduced pxf.session.user property and JDBC conn pool qualifier

---

## 5.10.0 (12/02/2019)

#### Enhancements:

- [#248](https://github.com/greenplum-db/pxf/pull/268) Enable Avro write
- [#261](https://github.com/greenplum-db/pxf/pull/261) Add Support for impersonation per server
- [#257](https://github.com/greenplum-db/pxf/pull/257) Add Support for Kerberized Hive 3
- [#254](https://github.com/greenplum-db/pxf/pull/254) Merge pxf impersonation jdbc
- [#247](https://github.com/greenplum-db/pxf/pull/247) Add support for multiple kerberos Hadoop and Hive servers

#### Bug Fixes:

- [#268](https://github.com/greenplum-db/pxf/pull/268) Fallback on using LineRecordReader when reading encrypted files

---

## 5.9.2 (10/28/2019)

#### Bug Fixes:

- [#253](https://github.com/greenplum-db/pxf/pull/253) Improve performance of HdfsFileFragmenter

---

## 5.9.1 (10/15/2019)

#### Bug Fixes:

- [#251](https://github.com/greenplum-db/pxf/pull/251) Fix regression in *:text:multi profiles when using wildcards

---

## 5.9.0 (09/25/2019)

#### Enhancements:

- [#243](https://github.com/greenplum-db/pxf/pull/243) Upgrade jackson libraries to 2.9.10
- [#235](https://github.com/greenplum-db/pxf/pull/235) Certify support for Hive 3.1
- [#226](https://github.com/greenplum-db/pxf/pull/226) Add OR and NOT support for JDBC filter pushdown
- [#236](https://github.com/greenplum-db/pxf/pull/236) Upgrade tomcat to version 7.0.96
- [#230](https://github.com/greenplum-db/pxf/pull/230) Add support for Hive 2 (Up to Hive 2.3.6)

#### Notes:

- PXF does not support Hive when running Java 11. As a workaround run PXF on Java 8.

---

## 5.8.2 (09/08/2019)

#### Bug Fixes:

- [#228](https://github.com/greenplum-db/pxf/pull/228) Make JDBC profile not fail when MAPR JAR files override default Hadoop ones
- [#217](https://github.com/greenplum-db/pxf/pull/217) CLI: reset on standby master and don't allow cluster init without PXF_CONF set #217
- [#224](https://github.com/greenplum-db/pxf/pull/224) Fix cloud access when Kerberized Hadoop is present

---

## 5.8.1 (08/22/2019)

#### Enhancements:

- Enable multinode testing against GCP dataproc. Run automation tests against Hadoop 2.9.2 and Hive 2.3.5

---

## 5.8.0 (08/01/2019)

#### Bug Fixes:

- [#211](https://github.com/greenplum-db/pxf/pull/211) Preserve error when re-throwing IOException (#211)

#### Enhancements:

- [#187](https://github.com/greenplum-db/pxf/pull/187) Implement S3 Select
- [#189](https://github.com/greenplum-db/pxf/pull/189) Implement cluster reset command
- [#191](https://github.com/greenplum-db/pxf/pull/191) Support config option to specify the server configuration directory (#191)
- [#198](https://github.com/greenplum-db/pxf/pull/198) Support serializing a list of OneFields to CSV (#198)
- [#201](https://github.com/greenplum-db/pxf/pull/201) Add JDK11 to PXF docker base dev image (#201)
- [#202](https://github.com/greenplum-db/pxf/pull/202) Support NOT and OR operators for S3 Select
- [#203](https://github.com/greenplum-db/pxf/pull/203) Support reading and writing of timestamp with time zone for Parquet (#203)
- [#206](https://github.com/greenplum-db/pxf/pull/206) Add S3 Select support for Parquet using S3-SELECT=AUTO (#206)
- [#207](https://github.com/greenplum-db/pxf/pull/207) Enable support for PXF server to run with Java-11 (#207)
- [#212](https://github.com/greenplum-db/pxf/pull/212) Use format options for S3 Select (#212)
- [#213](https://github.com/greenplum-db/pxf/pull/213) Rename S3-SELECT option -> S3_SELECT (#213)

---

## 5.7.0 (07/08/2019)

#### Bug Fixes:

- [#193](https://github.com/greenplum-db/pxf/pull/193) Fix uncompressed write parquet (#193)
- [#170](https://github.com/greenplum-db/pxf/pull/170) JDBC: Query data from ranges outside of partition range (#170)
- [#192](https://github.com/greenplum-db/pxf/pull/192) JsonResolver: throw BadRecordException on bad JSON

#### Enhancements:

- [#196](https://github.com/greenplum-db/pxf/pull/196) Purge codehaus from codebase, and replace it with fasterXML library (#196)

---

## 5.6.0 (06/19/2019)

#### Bug Fixes:

- [#182](https://github.com/greenplum-db/pxf/pull/182) Run a named query that ends with semicolon

#### Enhancements:

- [#188](https://github.com/greenplum-db/pxf/pull/188) Support for FDW
- [#186](https://github.com/greenplum-db/pxf/pull/186) Enable JDBC Connection Pooling
- [#183](https://github.com/greenplum-db/pxf/pull/183) Upgrade postgres driver to version 42.2.5

---

## 5.5.1 (06/03/2019)

#### Bug Fixes:

- [#180](https://github.com/greenplum-db/pxf/pull/180) Upgrade the Postgres JDBC Driver version

---

## 5.5.0 (05/31/2019)

#### Bug Fixes:

- [#176](https://github.com/greenplum-db/pxf/pull/176) Upgrade jackson 2 version from 2.9.8 -> 2.9.9

#### Enhancements:

- [#171](https://github.com/greenplum-db/pxf/pull/171) Enable JDBC connection to Hive and JDBC-specific user impersonation per server
- [#178](https://github.com/greenplum-db/pxf/pull/178) Make pxf threads configurable
- [#172](https://github.com/greenplum-db/pxf/pull/172) Support user-specific configuration in server
- [#169](https://github.com/greenplum-db/pxf/pull/169) HdfsFileFragmenter
- [#162](https://github.com/greenplum-db/pxf/pull/162) Kill JVM/Tomcat on OutOfMemoryError
- [#160](https://github.com/greenplum-db/pxf/pull/160) JDBC: Optimize resolver for INSERT queries

---

## 5.4.0 (05/13/2019)

#### Bug Fixes:

- [#161](https://github.com/greenplum-db/pxf/pull/161) Fix Oracle timestamp wrapping

#### Enhancements:

- [#157](https://github.com/greenplum-db/pxf/pull/157) JDBC profile can execute a query to read data from an external DB

concourse/scripts/pxf_common.bash
#!/bin/bash

GPHOME=/usr/local/greenplum-db-devel
PXF_HOME=${GPHOME}/pxf
MDD_VALUE=/data/gpdata/master/gpseg-1
PXF_COMMON_SRC_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)

# on purpose do not call this PXF_CONF so that it is not set during pxf operations
PXF_CONF_DIR=/home/gpadmin/pxf

JAVA_HOME=$(find /usr/lib/jvm -name 'java-1.8.0-openjdk*' | head -1)

if [[ -d gpdb_src/gpAux/extensions/pxf ]]; then
	PXF_EXTENSIONS_DIR=gpdb_src/gpAux/extensions/pxf
else
	PXF_EXTENSIONS_DIR=gpdb_src/gpcontrib/pxf
fi

function set_env() {
	export TERM=xterm-256color
	export TIMEFORMAT=$'\e[4;33mIt took %R seconds to complete this step\e[0m';
}

function run_regression_test() {
	ln -s "${PWD}/gpdb_src" ~gpadmin/gpdb_src
	cat > ~gpadmin/run_regression_test.sh <<-EOF
		#!/bin/bash
		source /opt/gcc_env.sh || true
		source ${GPHOME}/greenplum_path.sh
		source gpdb_src/gpAux/gpdemo/gpdemo-env.sh
		export PATH=\$PATH:${GPHD_ROOT}/bin

		cd "${PXF_EXTENSIONS_DIR}"
		make installcheck USE_PGXS=1

		[ -s regression.diffs ] && cat regression.diffs && exit 1

		exit 0
	EOF

	chown -R gpadmin:gpadmin "${PXF_EXTENSIONS_DIR}"
	chown gpadmin:gpadmin ~gpadmin/run_regression_test.sh
	chmod a+x ~gpadmin/run_regression_test.sh
	su gpadmin -c ~gpadmin/run_regression_test.sh
}

function build_install_gpdb() {

	# need a login shell to source /opt/gcc_env.sh (~root/.bashrc)
	bash --login -c "
		cd '${PWD}/gpdb_src' || return 1
		CC=\$(command -v gcc) CXX=\$(command -v g++) ./configure \
			--with-{perl,python,libxml,zstd} \
			--disable-{gpfdist,orca} \
			'--prefix=${GPHOME}'
		make -j4 -s
		make -s install
	"
}

function install_gpdb_binary() {
	if [[ -d bin_gpdb ]]; then
		mkdir -p ${GPHOME}
		tar -xzf bin_gpdb/*.tar.gz -C ${GPHOME}
	else
		build_install_gpdb
	fi

	local gphome python_dir python_version=2.7 export_pythonpath='export PYTHONPATH=$PYTHONPATH'
	if [[ ${TARGET_OS} == centos ]]; then
		# We can't use service sshd restart as service is not installed on CentOS 7.
		/usr/sbin/sshd &
		# CentOS 6 uses python 2.6
		if grep 'CentOS release 6' /etc/centos-release; then
			python_version=2.6
		fi
		python_dir=python${python_version}/site-packages
		export_pythonpath+=:/usr/lib/${python_dir}:/usr/lib64/$python_dir
	elif [[ ${TARGET_OS} == ubuntu ]]; then
		# Adjust GPHOME if the binary expects it to be /usr/local/gpdb
		gphome=$(grep ^GPHOME= /usr/local/greenplum-db-devel/greenplum_path.sh | cut -d= -f2)
		if [[ $gphome == /usr/local/gpdb ]]; then
			mv /usr/local/greenplum-db-devel /usr/local/gpdb
			GPHOME=/usr/local/gpdb
			PXF_HOME=${GPHOME}/pxf
		fi
		service ssh start
		python_dir=python${python_version}/dist-packages
		export_pythonpath+=:/usr/local/lib/$python_dir
	fi

	echo "$export_pythonpath" >> "${PXF_COMMON_SRC_DIR}/../../automation/tinc/main/tinc_env.sh"
}

function remote_access_to_gpdb() {
	# Copy cluster keys to root user
	passwd -u root
	cp -Rf cluster_env_files/.ssh/* /root/.ssh
	cp -f cluster_env_files/private_key.pem /root/.ssh/id_rsa
	cp -f cluster_env_files/public_key.pem /root/.ssh/id_rsa.pub
	cp -f cluster_env_files/public_key.openssh /root/.ssh/authorized_keys
	sed 's/edw0/hadoop/' cluster_env_files/etc_hostfile >> /etc/hosts
	# Copy cluster keys to gpadmin user
	rm -rf /home/gpadmin/.ssh/*
	cp cluster_env_files/.ssh/* /home/gpadmin/.ssh
	cp cluster_env_files/.ssh/*.pem /home/gpadmin/.ssh/id_rsa
	cp cluster_env_files/public_key.openssh /home/gpadmin/.ssh/authorized_keys
	{ ssh-keyscan localhost; ssh-keyscan 0.0.0.0; } >> /home/gpadmin/.ssh/known_hosts
	ssh "${SSH_OPTS[@]}" gpadmin@mdw "
		source ${GPHOME}/greenplum_path.sh &&
		export MASTER_DATA_DIRECTORY=${MDD_VALUE} &&
		echo 'host all all 10.0.0.0/16 trust' >> ${MDD_VALUE}/pg_hba.conf &&
		psql -d template1 <<-EOF && gpstop -u
			CREATE EXTENSION pxf;
			CREATE DATABASE gpadmin;
			CREATE ROLE root LOGIN;
		EOF
	"
}


function create_gpdb_cluster() {
	pushd gpdb_src/gpAux/gpdemo || return 1
	su gpadmin -c "source ${GPHOME}/greenplum_path.sh && make create-demo-cluster"
	popd || return 1
}

function add_remote_user_access_for_gpdb() {
	local username=${1}
	# load local cluster configuration
	pushd gpdb_src/gpAux/gpdemo || return 1

	echo "Adding access entry for ${username} to pg_hba.conf"
	su gpadmin -c "source ./gpdemo-env.sh; echo 'local    all     ${username}     trust' >> \${MASTER_DATA_DIRECTORY}/pg_hba.conf"

	echo "Restarting GPDB for change to pg_hba.conf to take effect"
	su gpadmin -c "source ${GPHOME}/greenplum_path.sh; source ./gpdemo-env.sh; gpstop -u"
	popd || return 1
}

function setup_gpadmin_user() {

	# Don't create gpadmin user if already exists
	if ! id -u gpadmin; then
		groupadd -g 1000 gpadmin && useradd -u 1000 -g 1000 -M gpadmin
		echo "gpadmin  ALL=(ALL)	   NOPASSWD: ALL" > /etc/sudoers.d/gpadmin
		groupadd supergroup && usermod -a -G supergroup gpadmin
		mkdir -p ~gpadmin/.ssh
		ssh-keygen -t rsa -N "" -f ~gpadmin/.ssh/id_rsa
		cat /home/gpadmin/.ssh/id_rsa.pub >> ~gpadmin/.ssh/authorized_keys
		chmod 0600 /home/gpadmin/.ssh/authorized_keys
		{ ssh-keyscan localhost; ssh-keyscan 0.0.0.0; } >> ~gpadmin/.ssh/known_hosts
		chown -R gpadmin:gpadmin ${GPHOME} ~gpadmin/.ssh # don't chown cached dirs ~/.m2, etc.
		echo -e "password\npassword" | passwd gpadmin 2> /dev/null
	fi
	cat <<-EOF >> /etc/security/limits.d/gpadmin-limits.conf
		gpadmin soft core unlimited
		gpadmin soft nproc 131072
		gpadmin soft nofile 65536
	EOF
	echo "export JAVA_HOME=${JAVA_HOME}" >> ~gpadmin/.bashrc
	if [[ -d gpdb_src/gpAux/gpdemo ]]; then
		chown -R gpadmin:gpadmin gpdb_src/gpAux/gpdemo
	fi

	if grep -i ubuntu /etc/os-release; then
		echo '[[ -f ~/.bashrc ]] && . ~/.bashrc' >> ~gpadmin/.bash_profile
		chown gpadmin:gpadmin ~gpadmin/.bash_profile
	fi
}

function install_pxf_client() {
	# recompile pxf.so file for dev environments only
	if [[ ${TEST_ENV} == dev ]]; then
		# need a login shell to source /opt/gcc_env.sh (~root/.bashrc)
		bash --login -c "
			source '${GPHOME}/greenplum_path.sh'
			USE_PGXS=1 make -C '${PXF_EXTENSIONS_DIR}' install
		"
	fi
}

function install_pxf_server() {
	if [[ ! -d ${PXF_HOME} ]]; then
		if [[ -d pxf_tarball ]]; then
			tar -xzf pxf_tarball/pxf.tar.gz -C ${GPHOME}
		else
			# requires login shell so that Go's dep is on PATH
			bash --login -c "
				export BUILD_NUMBER=${TARGET_OS} JAVA_TOOL_OPTIONS=-Dfile.encoding=UTF8
				make -C '${PWD}/pxf_src' install
			"
		fi
	fi
	chown -R gpadmin:gpadmin "${PXF_HOME}"
}

function setup_impersonation() {
	local GPHD_ROOT=${1}

	# enable impersonation by gpadmin user
	if [[ ${IMPERSONATION} == true ]]; then
		echo 'Impersonation is enabled, adding support for gpadmin proxy user'
		cat > proxy-config.xml <<-EOF
			<property>
			  <name>hadoop.proxyuser.gpadmin.hosts</name>
			  <value>*</value>
			</property>
			<property>
			  <name>hadoop.proxyuser.gpadmin.groups</name>
			  <value>*</value>
			</property>
			<property>
			  <name>hadoop.security.authorization</name>
			  <value>true</value>
			</property>
			<property>
			  <name>hbase.security.authorization</name>
			  <value>true</value>
			</property>
			<property>
			  <name>hbase.rpc.protection</name>
			  <value>authentication</value>
			</property>
			<property>
			  <name>hbase.coprocessor.master.classes</name>
			  <value>org.apache.hadoop.hbase.security.access.AccessController</value>
			</property>
			<property>
			  <name>hbase.coprocessor.region.classes</name>
			  <value>org.apache.hadoop.hbase.security.access.AccessController,org.apache.hadoop.hbase.security.access.SecureBulkLoadEndpoint</value>
			</property>
			<property>
			  <name>hbase.coprocessor.regionserver.classes</name>
			  <value>org.apache.hadoop.hbase.security.access.AccessController</value>
			</property>
		EOF
		sed -i -e '/<configuration>/r proxy-config.xml' "${GPHD_ROOT}/hadoop/etc/hadoop/core-site.xml" "${GPHD_ROOT}/hbase/conf/hbase-site.xml"
		rm proxy-config.xml
	elif [[ ${IMPERSONATION} == false ]]; then
		echo 'Impersonation is disabled, no proxy user setup performed.'
	else
		echo "ERROR: Invalid or missing CI property value: IMPERSONATION=${IMPERSONATION}"
		exit 1
	fi
	if ! find "${GPHD_ROOT}/hbase/lib" -name 'pxf-hbase-*.jar' | grep pxf-hbase; then
		cp "${PXF_HOME}"/lib/pxf-hbase-*.jar "${GPHD_ROOT}/hbase/lib"
	fi
}

function start_hadoop_services() {
	local GPHD_ROOT=${1}

	# Start all hadoop services
	"${GPHD_ROOT}/bin/init-gphd.sh"
	"${GPHD_ROOT}/bin/start-hdfs.sh"
	"${GPHD_ROOT}/bin/start-zookeeper.sh" &
	"${GPHD_ROOT}/bin/start-yarn.sh" &
	"${GPHD_ROOT}/bin/start-hbase.sh" &
	"${GPHD_ROOT}/bin/start-hive.sh" &
	wait
	export PATH=$PATH:${GPHD_ROOT}/bin

	# list running Hadoop daemons
	jps

	# grant gpadmin user admin privilege for feature tests to be able to run on secured cluster
	if [[ ${IMPERSONATION} == true ]]; then
		echo 'Granting gpadmin user admin privileges for HBase'
		echo "grant 'gpadmin', 'RWXCA'" | hbase shell
	fi
}

function init_and_configure_pxf_server() {
	echo 'Ensure pxf version can be run before pxf init'
	su gpadmin -c "${PXF_HOME}/bin/pxf version | grep -E '^PXF version [0-9]+.[0-9]+.[0-9]+$'" || exit 1

	echo 'Initializing PXF service'
	# requires a login shell to source startup scripts (JAVA_HOME)
	su --login -c "PXF_CONF=${PXF_CONF_DIR} ${PXF_HOME}/bin/pxf init" gpadmin

	# update impersonation value based on CI parameter
	if [[ ! ${IMPERSONATION} == true ]]; then
		echo 'Impersonation is disabled, updating pxf-env.sh property'
		su gpadmin -c "echo 'export PXF_USER_IMPERSONATION=false' >> ${PXF_CONF_DIR}/conf/pxf-env.sh"
	fi

	# update runtime JDK value based on CI parameter
	RUN_JDK_VERSION=${RUN_JDK_VERSION:-1.8}
	if [[ $RUN_JDK_VERSION == 11 ]]; then
		echo 'JDK 11 requested for runtime, setting PXF JAVA_HOME=/usr/lib/jvm/jdk-11 in pxf-env.sh'
		su gpadmin -c "echo 'export JAVA_HOME=/usr/lib/jvm/jdk-11' >> ${PXF_CONF_DIR}/conf/pxf-env.sh"
	fi
}

function configure_pxf_s3_server() {
	mkdir -p "${PXF_CONF_DIR}/servers/s3"
	cp "${PXF_CONF_DIR}/templates/s3-site.xml" "${PXF_CONF_DIR}/servers/s3/s3-site.xml"
	sed -i  -e "s|YOUR_AWS_ACCESS_KEY_ID|${ACCESS_KEY_ID}|" \
		-e "s|YOUR_AWS_SECRET_ACCESS_KEY|${SECRET_ACCESS_KEY}|" \
		"${PXF_CONF_DIR}/servers/s3/s3-site.xml"

	mkdir -p "${PXF_CONF_DIR}/servers/s3-invalid"
	cp "${PXF_CONF_DIR}/templates/s3-site.xml" "${PXF_CONF_DIR}/servers/s3-invalid/s3-site.xml"
}

function configure_pxf_default_server() {
	# copy hadoop config files to PXF_CONF_DIR/servers/default
	if [[ -d /etc/hadoop/conf/ ]]; then
		cp /etc/hadoop/conf/*-site.xml "${PXF_CONF_DIR}/servers/default"
	fi
	if [[ -d /etc/hive/conf/ ]]; then
		cp /etc/hive/conf/*-site.xml "${PXF_CONF_DIR}/servers/default"
	fi
	if [[ -d /etc/hbase/conf/ ]]; then
		cp /etc/hbase/conf/*-site.xml "${PXF_CONF_DIR}/servers/default"
	fi
}

function start_pxf_server() {
	# Check if some other process is listening on 5888
	netstat -tlpna | grep 5888 || true

	echo 'Starting PXF service'
	su --login -c "${PXF_HOME}/bin/pxf start" gpadmin
	# grep with regex to avoid catching grep process itself
	ps -aef | grep '[t]omcat'
}

function setup_minio() {
	echo 'Adding test bucket gpdb-ud-scratch to Minio ...'
	mkdir -p /opt/minio/data/gpdb-ud-scratch

	export MINIO_ACCESS_KEY=admin
	export MINIO_SECRET_KEY=password
	echo "Minio credentials: accessKey=${MINIO_ACCESS_KEY} secretKey=${MINIO_SECRET_KEY}"

	echo 'Starting Minio ...'
	/opt/minio/bin/minio server /opt/minio/data &

	# export minio credentials as access environment variables
	export ACCESS_KEY_ID=${MINIO_ACCESS_KEY}
	export SECRET_ACCESS_KEY=${MINIO_SECRET_KEY}
}
